File: fdd/backend/app/schemas/mapping.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.account_mapping import MappingConfidence, MappingStatus
from app.models.standard_line_item import FinancialStatement, LineItemCategory
from app.models.tie_out import TieOutStatus

# ── StandardLineItem ─────────────────────────────────────


class StandardLineItemRead(BaseModel):
    id: uuid.UUID
    code: str
    name_en: str
    name_ko: str
    category: LineItemCategory
    statement_type: FinancialStatement
    display_order: int
    parent_code: str | None
    is_subtotal: bool
    keywords: list[str] | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── AccountMapping ───────────────────────────────────────


class AccountMappingCreate(BaseModel):
    source_account_code: str = Field(..., min_length=1, max_length=50)
    source_account_name: str = Field(..., min_length=1, max_length=500)
    target_line_item_code: str = Field(..., min_length=1, max_length=50)
    confidence: MappingConfidence
    status: MappingStatus = MappingStatus.PROPOSED
    match_score: Decimal | None = None
    algorithm: str | None = None
    affected_amount: Decimal = Field(..., description="해당 계정 TB 잔액")


class AccountMappingUpdate(BaseModel):
    target_line_item_code: str | None = Field(default=None, max_length=50)
    status: MappingStatus | None = None
    rejection_reason: str | None = None


class AccountMappingApprove(BaseModel):
    approved_by: str = Field(..., min_length=1, max_length=100)


class AccountMappingRead(BaseModel):
    id: uuid.UUID
    deal_id: uuid.UUID
    source_account_code: str
    source_account_name: str
    target_line_item_code: str
    confidence: MappingConfidence
    status: MappingStatus
    match_score: Decimal | None
    algorithm: str | None
    affected_amount: Decimal
    approved_by: str | None
    approved_at: datetime | None
    rejection_reason: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MappingBulkCreate(BaseModel):
    """매핑 일괄 저장 요청."""

    mappings: list[AccountMappingCreate] = Field(..., min_length=1)


# ── MappingSuggestion (매핑 제안 결과 DTO) ────────────────


class MappingSuggestion(BaseModel):
    """매핑 엔진이 반환하는 제안 결과 (DB 저장 전)."""

    source_account_code: str
    source_account_name: str
    suggested_target_code: str
    suggested_target_name_en: str
    suggested_target_name_ko: str
    confidence: MappingConfidence
    match_score: Decimal
    algorithm: str
    affected_amount: Decimal


# ── TieOutResult ─────────────────────────────────────────


class TieOutResultRead(BaseModel):
    id: uuid.UUID
    deal_id: uuid.UUID
    snapshot_id: uuid.UUID
    statement_type: FinancialStatement
    status: TieOutStatus
    tb_total: Decimal
    reconstructed_total: Decimal
    variance: Decimal
    variance_percentage: Decimal
    unmapped_account_count: int
    unmapped_total: Decimal
    top_discrepancies: list[dict] | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TieOutRun(BaseModel):
    """Tie-out 실행 요청."""

    snapshot_id: uuid.UUID

File: fdd/frontend/src/types/mapping.ts
// ── Enums ────────────────────────────────────────────────

export type MappingConfidence = "HIGH" | "MEDIUM" | "LOW" | "UNMAPPED";
export type MappingStatus = "PROPOSED" | "APPROVED" | "REJECTED" | "MANUAL";
export type TieOutStatus = "PASS" | "FAIL" | "WARNING";
export type FinancialStatement = "IS" | "BS";
export type LineItemCategory =
  | "REVENUE"
  | "COGS"
  | "SGA"
  | "OTHER_OPERATING_INCOME"
  | "DEPRECIATION_AMORTIZATION"
  | "NON_OPERATING"
  | "INTEREST_EXPENSE"
  | "INTEREST_INCOME"
  | "TAX_EXPENSE"
  | "CASH"
  | "AR"
  | "INVENTORY"
  | "OTHER_CURRENT_ASSETS"
  | "PPE"
  | "INTANGIBLES"
  | "OTHER_NONCURRENT_ASSETS"
  | "AP"
  | "ACCRUALS"
  | "OTHER_CURRENT_LIABILITIES"
  | "DEBT"
  | "LEASE_LIABILITIES"
  | "OTHER_NONCURRENT_LIABILITIES"
  | "EQUITY";

// ── Standard Line Item ───────────────────────────────────

export interface StandardLineItem {
  id: string;
  code: string;
  name_en: string;
  name_ko: string;
  category: LineItemCategory;
  statement_type: FinancialStatement;
  display_order: number;
  parent_code: string | null;
  is_subtotal: boolean;
  keywords: string[] | null;
  created_at: string;
}

// ── Account Mapping ──────────────────────────────────────

export interface AccountMappingRead {
  id: string;
  deal_id: string;
  source_account_code: string;
  source_account_name: string;
  target_line_item_code: string;
  confidence: MappingConfidence;
  status: MappingStatus;
  match_score: string | null;
  algorithm: string | null;
  affected_amount: string;
  approved_by: string | null;
  approved_at: string | null;
  rejection_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface AccountMappingCreate {
  source_account_code: string;
  source_account_name: string;
  target_line_item_code: string;
  confidence: MappingConfidence;
  status?: MappingStatus;
  match_score?: string | null;
  algorithm?: string | null;
  affected_amount: string;
}

export interface MappingBulkCreate {
  mappings: AccountMappingCreate[];
}

export interface AccountMappingUpdate {
  target_line_item_code?: string;
  status?: MappingStatus;
  rejection_reason?: string;
}

export interface AccountMappingApprove {
  approved_by: string;
}

// ── Mapping Suggestion ───────────────────────────────────

export interface MappingSuggestion {
  source_account_code: string;
  source_account_name: string;
  suggested_target_code: string;
  suggested_target_name_en: string;
  suggested_target_name_ko: string;
  confidence: MappingConfidence;
  match_score: string;
  algorithm: string;
  affected_amount: string;
}

// ── Tie-out ──────────────────────────────────────────────

export interface TieOutResultRead {
  id: string;
  deal_id: string;
  snapshot_id: string;
  statement_type: FinancialStatement;
  status: TieOutStatus;
  tb_total: string;
  reconstructed_total: string;
  variance: string;
  variance_percentage: string;
  unmapped_account_count: number;
  unmapped_total: string;
  top_discrepancies: Array<Record<string, string>> | null;
  created_at: string;
}

export interface TieOutRun {
  snapshot_id: string;
}

// ── Evidence Missing Detection ───────────────────────────

export interface MissingEvidenceItem {
  target_type: string;
  target_id: string;
  target_label: string;
  reason: string;
}

export interface DetectionResult {
  deal_id: string;
  total_targets_checked: number;
  missing_count: number;
  coverage_percentage: number;
  missing: MissingEvidenceItem[];
}
